### PR TITLE
State Of The Firmware #4

### PR DESCRIPTION
**Pull request by petejohanson**
Initial draft, to start getting eyes on this. Several sections fully missing details, a few needing updated links to the relevant PRs, etc.

@innovaker @Nicell @okke-formsma @joelspadin Any thing I've missed?

### CODE REVIEW

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+#### Grace-Escape
+
+[okke-formsma] added an implementation of the "Grave Escape" behavior, developing a more generic "mod-morph" behavior to do so. Adding
+
+```
+&grave
+```
+
+to your keymap will send `ESC` when pressed on its own, but will send `` ` `` when pressed with a modifier held.
```
Comment (okke-formsma) [inline docs/blog/2021-01-25-zmk-sotf-4.md:75]: with a gui or shift modifier held.

**Review by okke-formsma — COMMENTED** [review]
I have only a tiny improvement for the gresc text. Thanks for the nice words :)

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+[petejohanson] implemented the first two complete, dynamic "widgets" for the displays for ZMK, adding a small battery indicator, which includes charging status, and a small output indicator, showing the currently active output (USB or BLE). When using BLE, the indicator also shows the active profile slot, as well as if the profile slot is open, awaiting connection from the paired host, or is actively connected to the host for that profile slot.
+
+#### Highest Layer Display
+
+[mcrosson] has contributed the next display widget, showing the highest active layer in the keymap. [petejohanson] then added a small follow up to allow layers in keymaps to add a `label` property to each layer, e.g. `label = "Nav";` and have that label be displayed in the widget instead of the numeric layer number.
+
+#### WPM
+
+New contributor [allymparker](https://github.com/allymparker) added our forth widget, a words-per-minute display! This widget work also included creating the core state logic for tracking the WPM.
```
Comment (joelspadin) [inline docs/blog/2021-01-25-zmk-sotf-4.md:99]:
```suggestion
New contributor [allymparker](https://github.com/allymparker) added our fourth widget, a words-per-minute display! This widget work also included creating the core state logic for tracking the WPM.
```

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+#### `&to` Layer Behavior
+
+[mcrosson] contributed the new [`&to`](/docs/behaviors/layers#to-layer) layer related behavior. This can be used to completely replace the active layer with a new one.
+
+This is most frequently used when using multiple core base layers with different layouts, e.g. QWERTY and DVORAK, to switch between them.
+
+#### Grace-Escape
+
+[okke-formsma] added an implementation of the "Grave Escape" behavior, developing a more generic "mod-morph" behavior to do so. Adding
```
Comment (joelspadin) [inline docs/blog/2021-01-25-zmk-sotf-4.md:69]: Should we link to documentation on mod-morph here? (Is there documentation for mod-morph/grave escape? I didn't find anything when I searched.)

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+- Other hosting costs, e.g. Docker Hub.
+
+For anyone looking to contribute, you can find the [ZMK Firmware project](https://opencollective.com/zmkfirmware) is now set up on [Open Collective](https://opencollective.com).
+
+## Coming Soon!
+
+Some items listed in the last coming soon section are still under active development.
+
+- A power profiler page for the website, to help users estimate their battery life for a given keyboard - [Nicell]
+- A keymap converter to automatically update keymaps to the new codes and use of `&kp` everywhere - [joelspadin]
```
Comment (joelspadin) [inline docs/blog/2021-01-25-zmk-sotf-4.md:165]: This one was merged a while ago! https://zmkfirmware.dev/docs/codes/keymap-upgrader

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 3 above)
Comment (mcrosson) [reply docs/blog/2021-01-25-zmk-sotf-4.md:69]: Agreed, I actually implemented a tweaked version of the grave-esc for my personal keymaps that does more of what i wanted from grave-esc than the defaults.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 2 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md]: Indeed! Good catch, thanks.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 4 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md]: Yup. Added as a section to the keymap part of the post, with link to the upgrader for folks to use.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 3 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md:69]: No docs for this yet, sadly, which is why I added a brief intro on how to use it here.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 1 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md]: Thanks! Updated.

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+for a sticky layer.
+
+#### `&to` Layer Behavior
+
+[mcrosson] contributed the new [`&to`](/docs/behaviors/layers#to-layer) layer related behavior. This can be used to completely replace the active layer with a new one.
+
+This is most frequently used when using multiple core base layers with different layouts, e.g. QWERTY and DVORAK, to switch between them.
+
+#### Grace Escape Behavior
```
Comment (Lenbok) [inline docs/blog/2021-01-25-zmk-sotf-4.md:67]: Typo: "Grace" -> "Grave"

File: docs/blog/2021-01-25-zmk-sotf-4.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+[petejohanson] added idle blanking for displays, which ensures they will go blank, and into low power mode, after a short period of inactivity from the user. This ensures we avoid burn-in for OLEDs, and helps improve battery life.
+
+#### Battery and Output Widgets
+
+[petejohanson] implemented the first two complete, dynamic "widgets" for the displays for ZMK, adding a small battery indicator, which includes charging status, and a small output indicator, showing the currently active output (USB or BLE). When using BLE, the indicator also shows the active profile slot, as well as if the profile slot is open, awaiting connection from the paired host, or is actively connected to the host for that profile slot.
+
+#### Highest Layer Display
+
+[mcrosson] has contributed the next display widget, showing the highest active layer in the keymap. [petejohanson] then added a small follow up to allow layers in keymaps to add a `label` property to each layer, e.g. `label = "Nav";` and have that label be displayed in the widget instead of the numeric layer number.
```
Comment (Lenbok) [inline docs/blog/2021-01-25-zmk-sotf-4.md:101]: Are there any keymaps in the repo that demonstrate the use of layers with labels?

Comment (mcrosson) [reply docs/blog/2021-01-25-zmk-sotf-4.md:101]: I don't believe so but it's very straight forward. See [here](https://git.kemonine.info/kemonine/keyboard/src/branch/master/zmk-config/config/nibble.keymap) if you'd like an example.

With the OLED not enabled by default I'm not sure it's necessary to say more than what's already here.

It'll be good to include in the OLED docs when they are written but for now there aren't any docs on OLED usage.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 5 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md:67]: Oy! Fixed. Thanks.

File: docs/blog/2021-01-25-zmk-sotf-4.md (hunk 6 above)
Comment (petejohanson) [reply docs/blog/2021-01-25-zmk-sotf-4.md:101]: Not yet. Documenting all of this in details is definitely part of the work remaining to finish this.

**Pull request merged**